scikits/odes/sundials/cvode.pyx

```diff
@@ -1180,22 +1205,66 @@ cdef class CVODE:
         np_tspan = np.asarray(tspan, dtype=float)
         np_y0    = np.asarray(y0, dtype=float)
 
-        return self._solve(np_tspan, np_y0)
+        soln = self._solve(np_tspan, np_y0)
+
+        flag_no, t, y, t_err, y_err = soln[:5]
+
+        if soln[5]:
+            t_roots = np.array(soln[5])
+        else:
+            t_roots = None
+
+        if soln[6]:
+            y_roots = np.array(soln[6])
+        else:
+            y_roots = None
+
+        if soln[7]:
+            t_tstop = np.array(soln[7])
+        else:
+            t_tstop = None
+
+        if soln[8]:
+            y_tstop = np.array(soln[8])
+        else:
+            y_tstop = None
+
+        flag = self.RETURN_FLAGS.get(flag_no, "UNKNOWN")
+
+        soln = CVODE_Return(
+            flag=flag, flag_no=flag_no, values=(t,y), err_values=(t_err,y_err),
+            root_values=(t_roots,y_roots), tstop_values=(t_tstop,y_tstop)
+        )
+
+        if self._old_api:
+            return flag_no, t, y, t_err, y_err
+        return soln
 
     cpdef _solve(self, np.ndarray[DTYPE_t, ndim=1] tspan,
                  np.ndarray[DTYPE_t, ndim=1] y0):
 
-        cdef np.ndarray[DTYPE_t, ndim=1] t_retn, t_interr
-        cdef np.ndarray[DTYPE_t, ndim=2] y_retn, y_interr
+        cdef np.ndarray[DTYPE_t, ndim=1] t_retn
+        cdef np.ndarray[DTYPE_t, ndim=2] y_retn
         t_retn  = np.empty(np.shape(tspan), float)
         y_retn  = np.empty([np.alen(tspan), np.alen(y0)], float)
 
+        cdef list t_roots
+        cdef list y_roots
+        t_roots = []
+        y_roots = []
+
+        cdef np.ndarray[DTYPE_t, ndim=1] y_tstop
+        cdef DTYPE_t t_tstop
+        y_tstop = np.empty(np.alen(y0), float)
+
```

Comment:
Well, tstop and root I intended to pay attention after all this new API is settled, but since you started with it... :-D
First, I was thinking about separating the 'interruptfn' into e.g. 'onroot' and 'ontstop' functions instead of having just one - I find this more intuitive in case one does not need both root or tstop - sure, no problem to ignore the extra argument. Opinions welcome.
Second, 'tstop' is also settable at runtime, so one can prolong the value in 'interruptfn' and continue the computation.
Question is whether it would be welcome to collect also tstop points. I believe it is useless (but can be added at anytime later if needed).

And, I like lists, so yeah, better :-D